### PR TITLE
Create backing file of arbitrary size for disk image

### PR DESCRIPTION
**Pull request by mloviska**
For bootable disk image testing in SLEM,MicroOS or JeOS we need to
create a disk image of different size than the image itself is shipped.
For QEMU backend this works already perfectly as this is handled well in
`OpenQA/Qemu` code.
The previous [attempt](https://github.com/os-autoinst/os-autoinst/pull/1717) lead to openQA test run failures in loading [pre-installed systems](https://openqa.suse.de/tests/6408034#step/boot_to_desktop/9)

- [Original issue in slem@s390x z/KVM](https://openqa.suse.de/tests/6590998#step/image_checks/7)
- VRs:
    * [sle-15-SP4-Online-s390x-Build18.1-create_hdd_textmode@s390x-kvm-sle12](http://kepler.suse.cz/tests/6274)
    * [sle-15-SP4-Online-s390x-Build18.1-qemu@s390x-kvm-sle12](http://kepler.suse.cz/tests/6276)
    * [slem image-checks](http://kepler.suse.cz/tests/6268#step/image_checks/4)
    * [JeOS with HDDSIZEGB_1=100](http://kepler.suse.cz/tests/6259)
    * [JeOS with HDDSIZEGB_1=undef](http://kepler.suse.cz/tests/6258)

### CODE REVIEW

**Comment by mloviska** [comment]
Hello @Soulofdestiny @jlausuch @Vogtinator, you have a look, pls?

**Review by okurz — CHANGES_REQUESTED** [review]
Please add tests.

Also I consider the function add_disk very complicated already. I suggest to extract some parts before adding more.

File: consoles/sshVirtsh.pm [hunk 1]
```diff
@@ -417,7 +418,17 @@ sub add_disk {
             }
             else {
                 $file = $basedir . $file;
-                $self->run_cmd(sprintf("qemu-img create '${file}' -f qcow2 -b '$basedir/%s'", $file_basename))
+                # requested expected value in GBs, if there is any passed as an argument
+                my $size = $args->{size} // 0;
+                # expected value in Bytes
+                my (undef, $json) = $self->run_cmd("qemu-img info --output=json $args->{file}", wantarray => 1);
+                my $image_vsize = decode_json($json)->{'virtual-size'};
+                if (($size * 1024 * 1024 * 1024) <= $image_vsize) {
```
Comment (jlausuch) [inline consoles/sshVirtsh.pm:426]:
```suggestion
                $size = (($size * 1024 * 1024 * 1024) <= $image_vsize) ? $image_vsize : $size . 'G'
```

if you prefer 1 liner, but ok with current implementation :)

**Comment by jlausuch** [comment]
LGTM, but I guess you need to add the unit tests.

File: consoles/sshVirtsh.pm [hunk 2]
```diff
@@ -417,7 +418,17 @@ sub add_disk {
...
             else {
                 $file = $basedir . $file;
-                $self->run_cmd(sprintf("qemu-img create '${file}' -f qcow2 -b '$basedir/%s'", $file_basename))
+                # requested expected value in GBs, if there is any passed as an argument
+                my $size = $args->{size} // 0;
+                # expected value in Bytes
+                my (undef, $json) = $self->run_cmd("qemu-img info --output=json $args->{file}", wantarray => 1);
+                my $image_vsize = decode_json($json)->{'virtual-size'};
+                $size = (($size * 1024 * 1024 * 1024) <= $image_vsize) ? $image_vsize : $size . 'G'
+                    $size = $image_vsize;
```
Comment (jlausuch) [inline consoles/sshVirtsh.pm:427]: I guess you don´t need these lines any more.

File: consoles/sshVirtsh.pm [hunk 3]
```diff
@@ -417,7 +418,17 @@ sub add_disk {
             }
             else {
                 $file = $basedir . $file;
-                $self->run_cmd(sprintf("qemu-img create '${file}' -f qcow2 -b '$basedir/%s'", $file_basename))
+                # requested expected value in GBs, if there is any passed as an argument
+                my $size = $args->{size} // 0;
+                # expected value in Bytes
+                my (undef, $json) = $self->run_cmd("qemu-img info --output=json $args->{file}", wantarray => 1);
+                my $image_vsize = decode_json($json)->{'virtual-size'};
+                $size = (($size * 1024 * 1024 * 1024) <= $image_vsize) ? $image_vsize : $size . 'G'
```
Comment (jlausuch) [inline consoles/sshVirtsh.pm:426]:
```suggestion
                $size = (($size * 1024 * 1024 * 1024) <= $image_vsize) ? $image_vsize : $size . 'G';
```

Comment (mloviska) [reply consoles/sshVirtsh.pm]: I have rebased it, thanks!

**Review by jlausuch — COMMENTED** [review]
LGTM but would like os-autoinst experts to review.

**Comment by okurz** [comment]
As you can see in the failed coverage check the patch coverage is too low. Meaning that there is no or too little test coverage of the changed lines of code. So https://github.com/os-autoinst/os-autoinst/pull/1736#pullrequestreview-718236676 still applies. If you state explicitly that you don't want to or can't do that then we can try to help or accept the change without as exception but for long-term sustainability I suggest to look into proper tests

**Comment by jlausuch** [comment]
> As you can see in the failed coverage check the patch coverage is too low. Meaning that there is no or too little test coverage of the changed lines of code. So [#1736 (review)](https://github.com/os-autoinst/os-autoinst/pull/1736#pullrequestreview-718236676) still applies. If you state explicitly that you don't want to or can't do that then we can try to help or accept the change without as exception but for long-term sustainability I suggest to look into proper tests

Since this is a minor change on something that was not even covered before (tests are not covering function `add_disk`), would it be ok to postpone the UT for later (a poo could be created for it) and merge this as it is? Anyway, with this PR, the coverage is just reduced 0.04% which is not very significant...

**Comment by jlausuch** [comment]
Thanks!

**Pull request merged**